### PR TITLE
Stripe processing fee calculation

### LINKED ISSUE #1014: Stripe Payment gateway Integration

### PR DESCRIPTION
**Pull request by ashu555**
Fixes #1014 

Run php artisan migrate -- path="database\migrations\2020_02_25_110853_add_stripe_api_key_to_api_keys_table.php"

### CODE REVIEW

**Comment by avinash403** [comment]
@ashu555 please put labels if requires review

File: app/Http/Controllers/Front/CheckoutController.php [hunk 1]
```diff
@@ -228,9 +228,12 @@ public function postCheckout(Request $request)
                     $attributes = $this->getAttributes($content);
                 }
             } else {
+                
                 $items = new \Illuminate\Support\Collection();
                 $invoiceid = $request->input('invoice_id');
                 $invoice = $this->invoice->find($invoiceid);
+                $processingFee = $payment_method =="razorpay"? 0 : $request->input('process_fee');
```
Comment (avinash403) [inline app/Http/Controllers/Front/CheckoutController.php:235]: `$request->input('process_fee')` is this coming from client panel?

File: app/Http/Controllers/Front/CheckoutController.php [hunk 2]
```diff
@@ -228,9 +228,12 @@ public function postCheckout(Request $request)
...
+                $invoice->grand_total = intval($invoice->grand_total+($invoice->grand_total*$processingFee/100)) ;
```
Comment (avinash403) [inline app/Http/Controllers/Front/CheckoutController.php:236]: can simplify it to `intval($invoice->grand_total*(1+$processingFee/100))`

File: app/Plugins/Stripe/Controllers/SettingsController.php [hunk 3]
```diff
@@ -182,28 +180,27 @@ public function postPaymentWithStripe(Request $request)
...
-                \Session::put('error','Money not add in wallet!!');
-                return redirect()->route('stripform');
+                \Session::put('fails','Your Payment was declined. Do you want to try making payment with the other gateway?');
+                return redirect()->route('checkout');
             }
         } catch (Exception $e) {
+            \Session::put('amount',$request['amount']);
             \Session::put('error',$e->getMessage());
             return redirect()->route('stripform');
         } catch(\Cartalyst\Stripe\Exception\CardErrorException $e) {
```
Comment (avinash403) [inline app/Plugins/Stripe/Controllers/SettingsController.php:198]: `CardErrorException` must be extending to `Exception`. In that case, it will never enter this block. Same goes for `MissingParameterException`

File: config/app.php [hunk 4]
```diff
@@ -152,7 +151,7 @@
             //
 
             // 'App\Plugins\Ccavanue\ServiceProvider',
-            'App\Plugins\Paypal\ServiceProvider',
+            // 'App\Plugins\Paypal\ServiceProvider',
```
Comment (avinash403) [inline config/app.php:154]: remove the line if not needed

File: resources/views/themes/default1/front/checkout.blade.php [hunk 5]
```diff
@@ -165,49 +165,51 @@
...
                 @if(count($gateways)>0 ) 
-                  <div class="form-group">
+                <div class="row">
+                  
 
                     <div class="col-md-6">
                         @foreach($gateways as $gateway)
-                        {{ucfirst($gateway)}} {!! Form::radio('payment_gateway',strtolower($gateway)) !!}<br><br>
+                        <?php
+                        $processingFee = \DB::table(strtolower($gateway))->where('currencies',$item['attributes']['currency']['currency'])->first()->processing_fee;
```
Comment (avinash403) [inline resources/views/themes/default1/front/checkout.blade.php:186]: `->first()` queries all the columns from the DB. If you need only `processing_fee`, you can use `->value('processing_fee')`method. It will query only that particular column

File: resources/views/themes/default1/front/checkout.blade.php [hunk 6]
```diff
@@ -165,49 +165,51 @@
...
             @endif
                 @if($rzpstatus ==1)
-                <div class="form-group">
-
-
-                   <div class="form-row">
+                <div class="row">
                     <div class="col-md-6">
-                        {!! Form::radio('payment_gateway',strtolower('Razorpay')) !!}
+                        <input type="radio" id="rzp_selected" data-currency=0 name='payment_gateway' value="razorpay"> &nbsp;&nbsp;&nbsp;
```
Comment (avinash403) [inline resources/views/themes/default1/front/checkout.blade.php:202]: any reason for not using `Form::radio`?

File: resources/views/themes/default1/front/checkout.blade.php [hunk 7]
```diff
@@ -165,49 +165,51 @@
...
+                <div class="row">
+                  
 
                     <div class="col-md-6">
                         @foreach($gateways as $gateway)
-                        {{ucfirst($gateway)}} {!! Form::radio('payment_gateway',strtolower($gateway)) !!}<br><br>
+                        <?php
+                        $processingFee = \DB::table(strtolower($gateway))->where('currencies',$item['attributes']['currency']['currency'])->first()->processing_fee;
+                        ?>
+                         <input type="radio"  data-currency="{{$processingFee}}" id="allow_gateway" name='payment_gateway'  value={{$gateway}}>
```
Comment (avinash403) [inline resources/views/themes/default1/front/checkout.blade.php:188]: please use `Form:radio` if functionality isn't complex

File: resources/views/themes/default1/front/checkout.blade.php [hunk 8]
```diff
@@ -434,4 +434,52 @@
...
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js" type="text/javascript"></script>
+<script>
+  $('#checkoutsubmitform').submit(function(){
+     $("#proceed").html("<i class='fa fa-circle-o-notch fa-spin fa-1x fa-fw'></i>Please Wait...")
+    $("#proceed").prop('disabled', true);
+
+  });
+     $(document).ready(function(){
+            var finalPrice = $('#total-price').val();
+            console.log(finalPrice);
```
Comment (avinash403) [inline resources/views/themes/default1/front/checkout.blade.php:446]: please remove if not needed

File: resources/views/themes/default1/front/checkout.blade.php [hunk 9]
```diff
@@ -434,4 +434,52 @@
...
+            var totalPrice = finalPrice;
+            $('#fee').hide();
+            $.ajax({
+                type:'POST',
+                data: {'processing_fee':processingFee,'price':totalPrice},
+                 beforeSend: function () {
+                 $('#response').html( "<img id='blur-bg' class='backgroundfadein' style='width: 50px; height:50 px; display: block; position:    fixed;' src='{!! asset('lb-faveo/media/images/gifloader3.gif') !!}'>");
+                },
+                 url: "{{url('update-final-price')}}",
+                 success: function () {
```
Comment (avinash403) [inline resources/views/themes/default1/front/checkout.blade.php:459]: remove this if we are not handling success or failure

File: resources/views/themes/default1/front/checkout.blade.php [hunk 10]
```diff
@@ -434,4 +434,52 @@
...
+            var totalPrice = finalPrice;
+            $('#fee').show();
+            $.ajax({
+                type:'POST',
+                data: {'processing_fee':processingFee,'price':totalPrice},
+                 beforeSend: function () {
+                 $('#response').html( "<img id='blur-bg' class='backgroundfadein' style='width: 50px; height:50 px; display: block; position:    fixed;' src='{!! asset('lb-faveo/media/images/gifloader3.gif') !!}'>");
+                },
+                 url: "{{url('update-final-price')}}",
+                 success: function () {
```
Comment (avinash403) [inline resources/views/themes/default1/front/checkout.blade.php:474]: remove this if we are not handling success or failure

File: resources/views/themes/default1/front/paynow.blade.php [hunk 11]
```diff
@@ -121,68 +121,47 @@
...
                        ?>
-                     @if($gateways) 
-                  <div class="form-group">
-
-                    <div class="col-md-6">
-                        @if(count($gateways)>0 ) 
-                  <div class="form-group">
+                    
+                
+                @if(count($gateways)>0 )
```
Comment (avinash403) [inline resources/views/themes/default1/front/paynow.blade.php:134]: `>0` isn't needed

File: resources/views/themes/default1/front/paynow.blade.php [hunk 12]
```diff
@@ -121,68 +121,47 @@
...
+                  <div class="row">
 
                     <div class="col-md-6">
                         @foreach($gateways as $gateway)
-                        {{ucfirst($gateway)}} {!! Form::radio('payment_gateway',strtolower($gateway)) !!}<br><br>
+                        <?php
+                          $processingFee = \DB::table(strtolower($gateway))->where('currencies',$invoice->currency)->value('processing_fee');
+                        ?>
+                        <input type="hidden" name="process_fee" value="{{$processingFee}}">
+                        <input type="radio"  data-currency="{{$processingFee}}" id="allow_gateway" name='payment_gateway'  value={{$gateway}}>
```
Comment (avinash403) [inline resources/views/themes/default1/front/paynow.blade.php:143]: please use `Form::radio` if functionality is simple enough

File: app/Http/Controllers/Front/CheckoutController.php (hunk 1 above)
Comment (ashu555) [reply app/Http/Controllers/Front/CheckoutController.php:235]: Getting value now in backend

File: app/Http/Controllers/Front/CheckoutController.php (hunk 2 above)
Comment (ashu555) [reply app/Http/Controllers/Front/CheckoutController.php:236]: Fixed

File: config/app.php (hunk 4 above)
Comment (ashu555) [reply config/app.php:154]: Fixed

File: resources/views/themes/default1/front/checkout.blade.php (hunk 5 above)
Comment (ashu555) [reply resources/views/themes/default1/front/checkout.blade.php:186]: Fixed

File: resources/views/themes/default1/front/checkout.blade.php (hunk 8 above)
Comment (ashu555) [reply resources/views/themes/default1/front/checkout.blade.php:446]: Fixed

File: resources/views/themes/default1/front/checkout.blade.php (hunk 9 above)
Comment (ashu555) [reply resources/views/themes/default1/front/checkout.blade.php:459]: Fixed

File: resources/views/themes/default1/front/checkout.blade.php (hunk 10 above)
Comment (ashu555) [reply resources/views/themes/default1/front/checkout.blade.php:474]: Fixed

File: resources/views/themes/default1/front/paynow.blade.php (hunk 11 above)
Comment (ashu555) [reply resources/views/themes/default1/front/paynow.blade.php:134]: Fixed

File: app/Plugins/Stripe/Controllers/SettingsController.php (hunk 3 above)
Comment (ashu555) [reply app/Plugins/Stripe/Controllers/SettingsController.php:198]: Fixed

File: resources/views/themes/default1/front/checkout.blade.php (hunk 6 above)
Comment (ashu555) [reply resources/views/themes/default1/front/checkout.blade.php:202]: Fixed

File: resources/views/themes/default1/front/checkout.blade.php (hunk 7 above)
Comment (ashu555) [reply resources/views/themes/default1/front/checkout.blade.php:188]: Fixed

File: resources/views/themes/default1/front/paynow.blade.php (hunk 12 above)
Comment (ashu555) [reply resources/views/themes/default1/front/paynow.blade.php:143]: Fixed

File: resources/views/themes/default1/front/checkout.blade.php [hunk 13]
```diff
@@ -434,4 +434,43 @@
...
+
+  });
+     $(document).ready(function(){
+            var finalPrice = $('#total-price').val();
+            $("#rzp_selected").click(function(){
+            
+            var processingFee = $(this).attr('data-currency');
+            var totalPrice = finalPrice;
+            $('#fee').hide();
+            $.ajax({
```
Comment (avinash403) [inline resources/views/themes/default1/front/checkout.blade.php:451]: do we not have a csrf check in our backend?

Comment (ashu555) [reply resources/views/themes/default1/front/checkout.blade.php:451]: added now

File: app/Http/Controllers/Front/CheckoutController.php [hunk 14]
```diff
@@ -228,9 +228,12 @@ public function postCheckout(Request $request)
                     $attributes = $this->getAttributes($content);
                 }
             } else {
+                
                 $items = new \Illuminate\Support\Collection();
                 $invoiceid = $request->input('invoice_id');
                 $invoice = $this->invoice->find($invoiceid);
+                $processingFee = $payment_method =="razorpay"? 0 : \DB::table(strtolower($request->input('payment_gateway')))->where('currencies',$invoice->currency)->value('processing_fee');
```
Comment (avinash403) [inline app/Http/Controllers/Front/CheckoutController.php:235]: not a good practice to directly use user input in table name. If this query was something like 
```
\DB::table($request-user_input))->get();
```
A lot of things could have been exploited

File: database/seeds/StripeSupportedCurrencySeeder.php [hunk 15]
```diff
@@ -14,807 +14,807 @@ class StripeSupportedCurrencySeeder extends Seeder
...
 		'processing_fee'               => '5',
 		]);
 
-             Stripe::create([
-		'supported_currencies'             => 'LKR',
+             StripePayment::create([
+		'currencies'             => 'LKR',
 		'base_currency'               => 'USD',
 		'processing_fee'               => '5',
 		]);
```
Comment (avinash403) [inline database/seeds/StripeSupportedCurrencySeeder.php:820]: Are we going to truncate the table and then run this seeder?

File: resources/views/themes/default1/front/checkout.blade.php [hunk 16]
```diff
@@ -165,49 +165,51 @@
...
                     <div class="col-md-6">
                         @foreach($gateways as $gateway)
-                        {{ucfirst($gateway)}} {!! Form::radio('payment_gateway',strtolower($gateway)) !!}<br><br>
+                        <?php
+                        $processingFee = \DB::table(strtolower($gateway))->where('currencies',$item['attributes']['currency']['currency'])->value('processing_fee');
+                        ?>
+                        {!! Form::radio('payment_gateway',$gateway,false,['id'=>'allow_gateway','data-currency'=>$processingFee]) !!}
+                         <img alt="{{$gateway}}" width="111"  src="{{asset('client/images/'.$gateway.'.png')}}">
+                          <br><br>
+                       <div id="fee" style="display:none"><p>An extra processing fee of <b>{{$processingFee}}%</b> will be charged on your Order Total during the time of payment</p></div>
```
Comment (avinash403) [inline resources/views/themes/default1/front/checkout.blade.php:191]: why is this div hidden?

File: resources/views/themes/default1/front/paynow.blade.php [hunk 17]
```diff
@@ -121,78 +121,56 @@
...
                     <div class="col-md-6">
                         @foreach($gateways as $gateway)
-                        {{ucfirst($gateway)}} {!! Form::radio('payment_gateway',strtolower($gateway)) !!}<br><br>
+                        <?php
+                          $processingFee = \DB::table(strtolower($gateway))->where('currencies',$invoice->currency)->value('processing_fee');
+                        ?>
+                        {!! Form::radio('payment_gateway',$gateway,false,['id'=>'allow_gateway']) !!}
+                         <img alt="Porto" width="111"  data-sticky-width="52" data-sticky-height="10" data-sticky-top="10" src="{{asset('client/images/'.$gateway.'.png')}}">
+                          <br><br>
+                         <div id="fee" style="display:none"><p>An extra processing fee of <b>{{$processingFee}}%</b> will be charged on your Order Total during the time of payment</p></div>
```
Comment (avinash403) [inline resources/views/themes/default1/front/paynow.blade.php:145]: why is this hidden?

File: database/seeds/StripeSupportedCurrencySeeder.php (hunk 15 above)
Comment (ashu555) [reply database/seeds/StripeSupportedCurrencySeeder.php:820]: No, we create the table and soon after seed these values

File: resources/views/themes/default1/front/checkout.blade.php (hunk 16 above)
Comment (ashu555) [reply resources/views/themes/default1/front/checkout.blade.php:191]: We only show this div when Stripe payment radio button is checked

File: resources/views/themes/default1/front/paynow.blade.php (hunk 17 above)
Comment (ashu555) [reply resources/views/themes/default1/front/paynow.blade.php:145]: We only show this div when Stripe payment radio button is checked

File: app/Http/Controllers/Front/CheckoutController.php (hunk 14 above)
Comment (ashu555) [reply app/Http/Controllers/Front/CheckoutController.php:235]: For now I have handled exception for SQL error. In future we can change Plugins DB Structure to handle such situation.

**Comment by SarfrazS** [comment]
Once you fix then I will Razorpay payment method also
